### PR TITLE
Added support for implicitly-typed out variables.

### PR DESCRIPTION
**Pull request by AlekseyTs**
Also addressed feedback provided for the previous change.

@dotnet/roslyn-compiler Please review.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/CSharpResources.resx [hunk 1]
```diff
@@ -4873,6 +4873,15 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
     <value>tuples</value>
   </data>
   <data name="IDS_FeatureOutVar" xml:space="preserve">
-    <value>out var</value>
+    <value>out variable declaration</value>
+  </data>
+  <data name="ERR_OutVariableUsedInTheSameArgumentList" xml:space="preserve">
+    <value>Reference to variable '{0}' is not permitted in the same argument list.</value>
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/CSharpResources.resx]: Nothing in the spec supports this error.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 2]
```diff
@@ -1369,12 +1369,42 @@ private BoundExpression BindNonMethod(SimpleNameSyntax node, Symbol symbol, Diag
...
+                                sourceLocal.IdentifierToken.Parent?.Kind() == SyntaxKind.Argument)
+                            {
+                                var argument = (ArgumentSyntax)sourceLocal.IdentifierToken.Parent;
+
+                                if (argument.Identifier == sourceLocal.IdentifierToken &&
+                                    argument.Type.IsVar)
+                                {
+                                    // We are referring to an out variable which might need type inference.
+                                    // If it is in fact needs inference, it is illegal to reference it in the same argument list. 
+                                    // Technically, we could support some restricted scenarios (when it is used as another argument in the same list),
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: You deleted this constraint from the draft specification.

File: src/Compilers/CSharp/Portable/Binder/PatternVariableFinder.cs [hunk 3]
```diff
@@ -165,6 +165,17 @@ public override void VisitArgument(ArgumentSyntax node)
                 return;
             }
 
+            switch (node.Identifier.Parent.Parent.Parent.Kind())
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/PatternVariableFinder.cs]: This expression "node.Identifier.Parent.Parent.Parent" is mysterious. A comment to explain would be helpful.
What happens if an "out var" is given as an indexer argument?

File: src/Compilers/CSharp/Portable/Binder/Binder.cs [hunk 4]
```diff
@@ -168,6 +168,18 @@ internal virtual ImmutableArray<LocalFunctionSymbol> GetDeclaredLocalFunctionsFo
             return this.Next.GetDeclaredLocalFunctionsForScope(scopeDesignator);
         }
 
+        /// <summary>
+        /// If this binder owns a scope for locals, return syntax node that is used
+        /// as the scope designator. Otherwise, null.
+        /// </summary>
+        internal virtual SyntaxNode ScopeDesignator
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/Binder.cs]: Does this correspond to the concept of scope designator as produced by the compiler in support of E&C?

File: src/Compilers/CSharp/Portable/BoundTree/OutVarLocalPendingInference.cs [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+        {
+            if (diagnosticsOpt != null)
+            {
+                Binder.Error(diagnosticsOpt, ErrorCode.ERR_TypeInferenceFailedForImplicitlyTypedOutVariable, ((ArgumentSyntax)this.Syntax).Identifier);
+            }
+
+            return this.SetInferredType(binder.CreateErrorType("var"), success: false);
+        }
+
+        public override object Display
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/BoundTree/OutVarLocalPendingInference.cs]: The convention in the compiler is that all the Display implementations are in Formatting.cs

File: src/Compilers/CSharp/Portable/BoundTree/OutVarLocalPendingInference.cs [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+        {
+            get
+            {
+                return CSharpResources.IDS_FeatureOutVar;
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/BoundTree/OutVarLocalPendingInference.cs]: .Localize()

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs [hunk 7]
```diff
@@ -519,10 +519,20 @@ public SymbolInfo GetSymbolInfo(ExpressionSyntax expression, CancellationToken c
                 // Named arguments handled in special way.
                 return this.GetNamedArgumentSymbolInfo((IdentifierNameSyntax)expression, cancellationToken);
             }
-            else
+            else if (expression.Parent?.Kind() == SyntaxKind.Argument &&
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs]: Rather than make a special case for this construct, I suggest you add a BoundTypeExpression to the declaration, so that the type would be found through the usual mechanism.

File: src/Compilers/CSharp/Portable/CSharpResources.resx (hunk 1 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/CSharpResources.resx]: Thanks, I'll add it. This error is reported only for an implicitly-typed error. I'll clarify the spec and adjust the message too.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 2 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: I deleted constraint for all out variables, this constraint is only for implicitly typed ones. I'll reflect this in the spec. Thanks.

File: src/Compilers/CSharp/Portable/Binder/Binder.cs (hunk 4 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder.cs]: I am not sure what are you referring to. This is a concept that binder uses internally in a couple of related APIs.

File: src/Compilers/CSharp/Portable/BoundTree/OutVarLocalPendingInference.cs (hunk 5 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/BoundTree/OutVarLocalPendingInference.cs]: Sure, I'll move it.

File: src/Compilers/CSharp/Portable/BoundTree/OutVarLocalPendingInference.cs (hunk 6 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/BoundTree/OutVarLocalPendingInference.cs]: Sure, I'll switch to localize infrastructure rather than reading the string from resources early.

File: src/Compilers/CSharp/Portable/Binder/PatternVariableFinder.cs (hunk 3 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/PatternVariableFinder.cs]: Added comments for each parent.

According to Language Specification, out/ref arguments are not allowed in the list of arguments for element access. Out variable declarations are disallowed in this context during the parsing.

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs (hunk 7 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs]: For correctness, the special case is needed only for implicitly-typed out variable, for other scenarios, existence of BoundTypeExpression would only be an optimization that prevents SemanticModel from binding the type syntax multiple time. Current implementation takes care of the optimization and handles implicitly-typed out variable at the same time. I considered adding BoundTypeExpression node, however, given that it only covers top level type name, adding another field into BoundLocal node, or introducing a new bound node altogether wouldn't pay for itself, I believe.

**Comment by AlekseyTs** [comment]
Addressed feedback provided so far.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 2 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: For out variables with a type specified, you do not have any constraints in the spec on where they may be accessed (though I believe you do implement some constraints). Please align the spec and implementation for those as well.

---
In reply to: [64161065](https://github.com/dotnet/roslyn/pull/11493#discussion_r64161065) [](ancestors = 64161065)

File: src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs [hunk 8]
```diff
@@ -606,6 +711,128 @@ static void Test2(object y, out int x)
...
+    }
+
+    static void Test2(object y, object x)
+    {
+    }
+}";
+            var compilation = CreateCompilationWithMscorlib(text, options: TestOptions.ReleaseExe, parseOptions: TestOptions.Regular.WithOutVarFeature());
+
+            compilation.VerifyDiagnostics(
+                // (6,15): error CS0841: Cannot use local variable 'x1' before it is declared
```
Comment (gafter) [inline src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs]: This error is not supported by the spec.

File: src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs [hunk 9]
```diff
@@ -606,6 +711,128 @@ static void Test2(object y, out int x)
...
+
+    static void Test2(out int y, object x)
+    {
+        y = 1;
+    }
+}";
+            var compilation = CreateCompilationWithMscorlib(text, options: TestOptions.ReleaseExe, parseOptions: TestOptions.Regular.WithOutVarFeature());
+
+            compilation.VerifyDiagnostics(
+                // (6,19): error CS0841: Cannot use local variable 'x1' before it is declared
```
Comment (gafter) [inline src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs]: Same comment, and many other places below.

File: src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs (hunk 8 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs]: Are you suggesting to specially mention this error situation in the outvar.md? Sure, I'll do that in the next PR.

---
In reply to: [64250908](https://github.com/dotnet/roslyn/pull/11493#discussion_r64250908) [](ancestors = 64250908)

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 2 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: Please be specific and point to the relevant places in this PR. Thanks.

---
In reply to: [64249559](https://github.com/dotnet/roslyn/pull/11493#discussion_r64249559) [](ancestors = 64249559,64161065)

File: src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs (hunk 9 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs]: Same response.

---
In reply to: [64252509](https://github.com/dotnet/roslyn/pull/11493#discussion_r64252509) [](ancestors = 64252509)

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review.

File: docs/features/outvar.md [hunk 10]
```diff
@@ -10,23 +10,19 @@ argument_value
...
 ```
 
-A variable declared this way is called an *out variable*. An *out variable* is read-only and scoped to the enclosing statement. More specifically, the scope will be the same as for a *pattern-variable* introduced via pattern-matching.
-
-> **Note**: We may treat *out variables* as *pattern variables* in the semantic model.
-
+A variable declared this way is called an *out variable*. 
 You may use the contextual keyword `var` for the variable's type.
+The scope will be the same as for a *pattern-variable* introduced via pattern-matching.
+Out variables are disallowed within constructor initializers.
```
Comment (gafter) [inline docs/features/outvar.md]: You said that this is an open issue. Please place it on the open issue list.

File: src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs (hunk 8 above)
Comment (gafter) [inline src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs]: I'm suggesting you deleted the only constraint on where the variable may be used, and then you implemented a constraint. The result is that the spec and implementation disagree (i.e. a bug). This can be fixed in one of four ways: fixing the spec, adding an issue in .work.md, opening an issue, or removing the constraint.

---
In reply to: [64252591](https://github.com/dotnet/roslyn/pull/11493#discussion_r64252591) [](ancestors = 64252591,64250908)

Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs]: Please be specific, what constraint did I remove (pleas quote it). Please also explain how this error is related to the constraint that is being removed.

---
In reply to: [64253737](https://github.com/dotnet/roslyn/pull/11493#discussion_r64253737) [](ancestors = 64253737,64252591,64250908)

Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs]: BTW, reporting of this error is not part of this PR. I simply added tests recommended for the previous PR.

---
In reply to: [64254529](https://github.com/dotnet/roslyn/pull/11493#discussion_r64254529) [](ancestors = 64254529,64253737,64252591,64250908)

Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs]: I have no idea what is this "comment recalled" business. It looks like CodeFlow is going crazy. 

---
In reply to: [64255531](https://github.com/dotnet/roslyn/pull/11493#discussion_r64255531) [](ancestors = 64255531,64255223,64254529,64253737,64252591,64250908)

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 2 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: The specific place in this PR is the deletion of the paragraph starting at line 21 of outvar.md in the previous revision of that file. To align the spec with the implementation, I recommend you describe a proposed change to the followin paragraph of the existing C# specification:

> Within the scope of a local variable introduced by a local-variable-declaration, it is a compile-time error to refer to that local variable in a textual position that precedes its local-variable-declarator. If the local variable declaration is implicit (§8.5.1), it is also an error to refer to the variable within its local-variable-declarator.

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: I think you are confused. This place in code doesn't implement the constraint that you are suggesting to document. It is, however, implements a constraint that is documented in outvar.md on line 19
```
It is illegal to reference implicitly-typed out variable in the same argument list. 
Technically, we could support some restricted scenarios (when it is used as another argument in the same list), 
but their utility is very questionable. Otherwise, we are looking into getting into a cycle trying to infer its type.
```

---
In reply to: [64256897](https://github.com/dotnet/roslyn/pull/11493#discussion_r64256897) [](ancestors = 64256897)

File: src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs [hunk 11]
```diff
@@ -809,6 +1076,1008 @@ static void Test(out System.Collections.Generic.IEnumerable<System.Int32> x)
...
+                // (7,30): error CS0103: The name 'x1' does not exist in the current context
+                //         Test2(x[out int x1], x1);
+                Diagnostic(ErrorCode.ERR_NameNotInContext, "x1").WithArguments("x1").WithLocation(7, 30)
+                );
+
+            var tree = compilation.SyntaxTrees.Single();
+            Assert.False(tree.GetRoot().DescendantNodes().OfType<ArgumentSyntax>().Where(p => p.Identifier.ValueText == "x1").Any());
+        }
+
         [Fact(Skip = "Needs adjustment. + No 'var' support yet.")]
```
Comment (gafter) [inline src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs]: You told me that "We have a policy, every skipped test should have a WorkItem attribute."

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 2 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: @AlekseyTs asked me to post this retraction and correction here.

I stand corrected. I was not precisely correct in my assertion.

The correct assertion is that the implementation resulting from this PR is erroneous, in that it results in a specification and implementation that do not agree with each other. Consequently the PR is not complete until that mismatch is either corrected or noted in one of the three ways approved by @jaredpar.

---
In reply to: [64259322](https://github.com/dotnet/roslyn/pull/11493#discussion_r64259322) [](ancestors = 64259322,64256897)

**Comment by AlekseyTs** [comment]
@gafter Please let me know when you are done with review pass. So that I could address all feedback at once. Thanks.

File: src/Compilers/CSharp/Portable/Binder/PatternVariableFinder.cs [hunk 12]
```diff
@@ -165,6 +165,21 @@ public override void VisitArgument(ArgumentSyntax node)
...
+                         Parent. // invocation/constructor initializer
+                         Kind())
+            {
+                case SyntaxKind.InvocationExpression:
+                case SyntaxKind.ObjectCreationExpression:
+                case SyntaxKind.ThisConstructorInitializer:
+                case SyntaxKind.BaseConstructorInitializer:
+                    break;
+                default:
+                    throw ExceptionUtilities.UnexpectedValue(node.Identifier.Parent.Parent.Parent.Kind());
```
Comment (VSadov) [inline src/Compilers/CSharp/Portable/Binder/PatternVariableFinder.cs]: It probably makes sense to have more detailed info for this exception. 
We are intentionally crashing here if this code ends up being reachable (because of compile errors or some unforeseen reasons). More detailed message might make it easier to diagnose the crash than just unexpected Kind.

I mean - if we do work to detect the case, it makes sense to make message a bit more detailed.

File: src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs (hunk 11 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs]: I think it should be Ok for a feature branch, but, if you insist, I'll remove the test.

---
In reply to: [64263247](https://github.com/dotnet/roslyn/pull/11493#discussion_r64263247) [](ancestors = 64263247)

**Comment by gafter** [comment]
@AlekseyTs I am done with this review pass.

File: src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs [hunk 13]
```diff
@@ -809,6 +1076,1008 @@ static void Test(out System.Collections.Generic.IEnumerable<System.Int32> x)
...
+        x = 125;
+        System.Console.WriteLine(x);
+        return null;
+    }
+
+    static object Test2(out short x, object y)
+    {
+        x = 126;
+        System.Console.WriteLine(x);
+        return null;
```
Comment (VSadov) [inline src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs]: Is there a test where both overloads have second arg of the same type? (error case)

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 14]
```diff
@@ -1369,12 +1369,42 @@ private BoundExpression BindNonMethod(SimpleNameSyntax node, Symbol symbol, Diag
                         }
                         else
                         {
+                            SourceLocalSymbol sourceLocal;
+                            type = null;
+
+                            if (node.SyntaxTree == localSymbolLocation.SourceTree && (object)(sourceLocal = localSymbol as SourceLocalSymbol) != null &&
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: What's the difference with the simpler check "sourceLocal is SourceLocalSymbol"?

Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: Nevermind, I got it :-)

---
In reply to: [64277268](https://github.com/dotnet/roslyn/pull/11493#discussion_r64277268) [](ancestors = 64277268)

File: src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs (hunk 11 above)
Comment (VSadov) [inline src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs]: "PROTOTYPE" in the Skip reason would also be acceptable.

File: src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs [hunk 15]
```diff
@@ -809,6 +1076,1008 @@ static void Test(out System.Collections.Generic.IEnumerable<System.Int32> x)
...
+                //         Test2(x[out int x1], x1);
+                Diagnostic(ErrorCode.ERR_NameNotInContext, "x1").WithArguments("x1").WithLocation(7, 25),
+                // (7,30): error CS0103: The name 'x1' does not exist in the current context
+                //         Test2(x[out int x1], x1);
+                Diagnostic(ErrorCode.ERR_NameNotInContext, "x1").WithArguments("x1").WithLocation(7, 30)
+                );
+
+            var tree = compilation.SyntaxTrees.Single();
+            Assert.False(tree.GetRoot().DescendantNodes().OfType<ArgumentSyntax>().Where(p => p.Identifier.ValueText == "x1").Any());
+        }
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs]: Possible test ideas: 

- `int x; M(out var x);` // should fail because x is already defined
- M(out var var);

**Comment by jcouv** [comment]
LGTM

File: src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs (hunk 15 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs]: Thanks, I have a work item that covers testing scope rules.

---
In reply to: [64279420](https://github.com/dotnet/roslyn/pull/11493#discussion_r64279420) [](ancestors = 64279420)

File: src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs (hunk 13 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs]: Good idea, I'll add one.

---
In reply to: [64276987](https://github.com/dotnet/roslyn/pull/11493#discussion_r64276987) [](ancestors = 64276987)

File: src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs (hunk 15 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/OutVarTests.cs]: I'll add one for M(out var var); 

---
In reply to: [64280745](https://github.com/dotnet/roslyn/pull/11493#discussion_r64280745) [](ancestors = 64280745,64279420)

**Comment by VSadov** [comment]
LGTM

File: src/Compilers/CSharp/Portable/Binder/PatternVariableFinder.cs (hunk 12 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/PatternVariableFinder.cs]: I looked through other places where we throw UnexpectedValue, not a single one of them uses more detailed message. I will add a comment instead.

---
In reply to: [64272964](https://github.com/dotnet/roslyn/pull/11493#discussion_r64272964) [](ancestors = 64272964)

Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/PatternVariableFinder.cs]: Consider extracting a local for `node.Identifier.Parent.Parent.Parent.Kind()`, if for no other reason so the two uses of that expression cannot get out of sync.

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/PatternVariableFinder.cs]: Sure.

---
In reply to: [64656212](https://github.com/dotnet/roslyn/pull/11493#discussion_r64656212) [](ancestors = 64656212)

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler I think I responded to all the feedback.

**Comment by AlekseyTs** [comment]
It looks like failure of windows_debug_unit64_prtest is a known issue https://github.com/dotnet/roslyn/issues/11368

**Comment by AlekseyTs** [comment]
I have more changes blocked on this PR. It has been more that 24 hours without new feedback and I do have two sign-offs. I am going to merge, but additional feedback is welcome.
